<a href="https://colab.research.google.com/github/ChronosXV/Aprendizaje-Automatico/blob/main/Aprendizaje_Automatico.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

** CONSUMO DIGITAL EN ADOLESCENTES**

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
df = pd.read_excel("/content/drive/MyDrive/Aprendizaje automatico/repositorio.xlsx")

**Analisis exploratorio**

In [ ]:
# Estado inicial del dataset
print("Filas:", df.shape[0])
print("Columnas:", df.shape[1])

Filas: 4115
Columnas: 125


Visualizamos el tipo de dato de cada columna, si tenemos filas duplicadas, columnas con el mismo nombre repetido

In [ ]:
print(df.dtypes.value_counts())
print(df.columns[df.columns.duplicated()]) # nombres de columna repetidos

float64           66
int64             37
object            21
datetime64[ns]     1
Name: count, dtype: int64
Index([], dtype='object')


No tenemos duplicados

In [ ]:
# vemos las distintas medidas
df.describe()

,Submission ID,Submission Date,investigador,Colegio,Ciudad,grado,edad,sexonum,Altura en m,Peso en kg,...,matematica,prom,aplazos,datoscompletoDEMOG,datoscompletoSUENO,datoscompletoPANTALLAS,excSUENOpantallas,excluidos,missingnotas,excluFINAL
count,4.115000e+03,4115,4115.000000,4115.000000,4115.000000,4115.000000,3553.000000,4115.000000,4081.000000,4056.000000,...,1257.000000,1257.000000,1257.000000,4115.000000,4115.000000,4115.000000,4115.000000,4115.000000,4115.000000,4115.000000
mean,4.407193e+18,2019-08-04 07:13:31.672903936,9.879708,43.297448,18.820656,3.215796,15.600338,0.446659,1.653658,57.782544,...,7.175123,7.308385,0.430390,0.863426,0.782503,0.869502,0.305225,0.404860,0.694532,0.790036
min,4.347980e+18,2019-05-27 18:26:20,1.000000,1.000000,0.000000,1.000000,12.000000,0.000000,1.400000,27.000000,...,1.000000,2.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,4.373565e+18,2019-06-26 09:08:15,3.000000,29.000000,14.000000,2.000000,14.000000,0.000000,1.600000,50.000000,...,6.000000,6.500000,0.000000,1.000000,1.000000,1.000000,0.000000,0.000000,0.000000,1.000000
50%,4.422135e+18,2019-08-21 14:17:34,7.000000,47.000000,16.000000,3.000000,16.000000,0.000000,1.650000,56.000000,...,7.000000,7.500000,0.000000,1.000000,1.000000,1.000000,0.000000,0.000000,1.000000,1.000000
75%,4.435872e+18,2019-09-06 11:53:09,15.000000,62.000000,27.000000,4.000000,17.000000,1.000000,1.720000,65.000000,...,9.000000,8.500000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000
max,4.494709e+18,2019-11-13 13:15:30,53.000000,89.000000,92.000000,6.000000,20.000000,1.000000,1.910000,90.000000,...,10.000000,10.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000
std,3.442841e+16,NaN,8.537678,22.669166,11.285266,1.518031,1.601527,0.497207,0.092749,11.320989,...,1.804322,1.440699,0.495328,0.343438,0.412594,0.336892,0.460559,0.490925,0.460661,0.407332


### Exploración de valores faltantes

Vamos a identificar las columnas con valores faltantes y el porcentaje de datos que faltan en cada una.

In [ ]:
# Buscamos valores nulos
print("Valores nulos por columna:")
print(df.isnull().sum())

Valores nulos por columna:
IP                        0
Submission ID             0
Submission Date           0
Código                    0
investigador              0
                         ..
datoscompletoPANTALLAS    0
excSUENOpantallas         0
excluidos                 0
missingnotas              0
excluFINAL                0
Length: 125, dtype: int64


In [ ]:
# Buscamos duplicados
cantidad_duplicados = df.duplicated().sum()

print(f"Cantidad de filas duplicadas: {cantidad_duplicados}")

Cantidad de filas duplicadas: 0


### Búsqueda de duplicados en columnas clave

Investigamos si existen duplicados basándonos en un subconjunto de columnas que deberían identificar entradas únicas, como 'IP' y 'Submission ID'.

In [ ]:
key_columns = ['IP', 'Submission ID']
duplicates_in_keys = df[df.duplicated(subset=key_columns, keep=False)]

print(f"Cantidad de filas duplicadas en '{', '.join(key_columns)}': {len(duplicates_in_keys)}")

if not duplicates_in_keys.empty:
    print("Primeras filas duplicadas:")
    display(duplicates_in_keys.sort_values(by=key_columns).head())
else:
    print(f"No se encontraron duplicados en las columnas {', '.join(key_columns)}.")

Cantidad de filas duplicadas en 'IP, Submission ID': 0
No se encontraron duplicados en las columnas IP, Submission ID.


### Revisión de valores faltantes

Cantidad de valores faltantes por columna

In [ ]:
# % de faltantes por columna, ordenado de mayor a menor
faltantes = df.isna().mean().sort_values(ascending=False) * 100
print(faltantes.head(30).round(2))

print()
# Los indicadores de completitud que ya trae el dataset
for col in ['datoscompletoDEMOG', 'datoscompletoSUENO', 'datoscompletoPANTALLAS', 'excluidos', 'excluFINAL']:
    if col in df.columns:
        print(col, '->', df[col].value_counts(dropna=False).to_dict())

WAZ                                              99.34
prom                                             69.45
aplazos                                          69.45
matematica                                       69.45
lengua                                           69.36
tpotvonlinePORC                                  27.85
tpotabletPORC                                    27.85
tporedesPORC                                     27.85
tpovideojuegosPORC                               27.85
tpootrasPORC                                     27.85
tpopantallatotalhstrimmed                        27.70
tpopantallatotalhscat                            27.70
tpopantallatotalhscat2                           27.70
BAZ                                              22.26
BMIzscoreCAT                                     22.26
HAZ                                              22.19
tposueñosemTRIMMED                               17.11
tposueñosemCAT                                   17.11
edad      

De esta forma detectamos que la columna WAZ tiene un porcentaje muy alto de valores faltantes. Por otro lado, las columnas prom, aplazos, matematica, lengua (~69% faltante) — el rendimiento académico está mayormente vacío

In [ ]:
flags = ['datoscompletoDEMOG', 'datoscompletoSUENO', 'datoscompletoPANTALLAS', 'excluFINAL']

# Cuántas filas quedan si exijo cada combinación
print('Solo excluFINAL==1:', (df['excluFINAL'] == 1).sum())
print('Solo pantallas completas:', (df['datoscompletoPANTALLAS'] == 1).sum())
print('Pantallas + sueño completos:', ((df['datoscompletoPANTALLAS'] == 1) & (df['datoscompletoSUENO'] == 1)).sum())
print('Los 3 bloques completos:', ((df['datoscompletoDEMOG'] == 1) & (df['datoscompletoSUENO'] == 1) & (df['datoscompletoPANTALLAS'] == 1)).sum())
print('Los 3 bloques + excluFINAL:', ((df['datoscompletoDEMOG'] == 1) & (df['datoscompletoSUENO'] == 1) & (df['datoscompletoPANTALLAS'] == 1) & (df['excluFINAL'] == 1)).sum())

# Edad: ¿el 13.66% de faltantes está dentro o fuera de excluFINAL==1?
print()
print('Edad faltante dentro de excluFINAL==1:', df.loc[df['excluFINAL'] == 1, 'edad'].isna().sum())
print('Edad faltante fuera de excluFINAL==1:', df.loc[df['excluFINAL'] == 0, 'edad'].isna().sum())

Solo excluFINAL==1: 3251
Solo pantallas completas: 3578
Pantallas + sueño completos: 2859
Los 3 bloques completos: 2449
Los 3 bloques + excluFINAL: 1748

Edad faltante dentro de excluFINAL==1: 399
Edad faltante fuera de excluFINAL==1: 163


Distintos criterios de filtro

In [ ]:
flags = ['datoscompletoDEMOG', 'datoscompletoSUENO', 'datoscompletoPANTALLAS', 'excluFINAL']

# Cuántas filas quedan si exijo cada combinación
print('Solo excluFINAL==1:', (df['excluFINAL'] == 1).sum())
print('Solo pantallas completas:', (df['datoscompletoPANTALLAS'] == 1).sum())
print('Pantallas + sueño completos:', ((df['datoscompletoPANTALLAS'] == 1) & (df['datoscompletoSUENO'] == 1)).sum())
print('Los 3 bloques completos:', ((df['datoscompletoDEMOG'] == 1) & (df['datoscompletoSUENO'] == 1) & (df['datoscompletoPANTALLAS'] == 1)).sum())
print('Los 3 bloques + excluFINAL:', ((df['datoscompletoDEMOG'] == 1) & (df['datoscompletoSUENO'] == 1) & (df['datoscompletoPANTALLAS'] == 1) & (df['excluFINAL'] == 1)).sum())

# Edad: ¿el 13.66% de faltantes está dentro o fuera de excluFINAL==1?
print()
print('Edad faltante dentro de excluFINAL==1:', df.loc[df['excluFINAL'] == 1, 'edad'].isna().sum())
print('Edad faltante fuera de excluFINAL==1:', df.loc[df['excluFINAL'] == 0, 'edad'].isna().sum())

Solo excluFINAL==1: 3251
Solo pantallas completas: 3578
Pantallas + sueño completos: 2859
Los 3 bloques completos: 2449
Los 3 bloques + excluFINAL: 1748

Edad faltante dentro de excluFINAL==1: 399
Edad faltante fuera de excluFINAL==1: 163


**Hallazgo importante**: dentro de excluFINAL == 1 (3.251 filas), igual falta edad en 399 casos (12,3%) — casi la misma proporción que fuera de ese filtro (163 de 864 = 18,9%). Eso quiere decir que excluFINAL no garantiza completitud demográfica. El flag de los autores se construyó mirando solo sueño y/o pantallas, no las variables demográficas que necesitamos  para caracterizar los clusters. Por lo tanto  hay que agregar el propio filtro de completitud demográfica.

In [ ]:
criterio = (
    (df['datoscompletoDEMOG'] == 1) &
    (df['datoscompletoSUENO'] == 1) &
    (df['datoscompletoPANTALLAS'] == 1) &
    (df['excluFINAL'] == 1)
)

df['grupo'] = criterio.map({True: 'incluido', False: 'excluido'})

print('Tamaños:')
print(df['grupo'].value_counts())

print()
print('=== Edad ===')
print(df.groupby('grupo')['edad'].describe()[['count','mean','std','min','max']].round(2))

print()
print('=== Sexo (%) ===')
print(pd.crosstab(df['grupo'], df['Sexo'], normalize='index').round(3) * 100)

print()
print('=== Tiempo total de pantalla (tpopantallatotalhstrimmed) ===')
print(df.groupby('grupo')['tpopantallatotalhstrimmed'].describe()[['count','mean','std','min','max']].round(2))

print()
print('=== Grado escolar ===')
print(df.groupby('grupo')['grado'].describe()[['count','mean','std','min','max']].round(2))

Tamaños:
grupo
excluido    2367
incluido    1748
Name: count, dtype: int64

=== Edad ===
           count   mean   std   min   max
grupo                                    
excluido  1805.0  15.75  1.60  12.0  20.0
incluido  1748.0  15.44  1.59  12.0  20.0

=== Sexo (%) ===
Sexo      Femenino  Masculino
grupo                        
excluido      59.3       40.7
incluido      50.0       50.0

=== Tiempo total de pantalla (tpopantallatotalhstrimmed) ===
           count  mean   std  min   max
grupo                                  
excluido  1581.0  6.75  2.95  0.0  13.0
incluido  1394.0  6.67  2.97  0.0  13.0

=== Grado escolar ===
           count  mean   std  min  max
grupo                                 
excluido  2367.0  3.40  1.51  1.0  6.0
incluido  1748.0  2.97  1.50  1.0  6.0


Conclusión: el criterio de filtro elegido no sesga la muestra en la variable central del trabajo (tiempo de pantalla) ni fuertemente en edad. Hay una diferencia de sexo entre quienes completan todos los datos y quienes no (las chicas parecen completar la encuesta de forma menos íntegra que los varones) "la muestra final tiene una composición de sexo más equilibrada (50/50) que la muestra completa, porque las respuestas incompletas se concentran algo más en participantes de sexo femenino".

Distribución de variables demográficas (sobre el subconjunto final)

In [ ]:
criterio = (
    (df['datoscompletoDEMOG'] == 1) &
    (df['datoscompletoSUENO'] == 1) &
    (df['datoscompletoPANTALLAS'] == 1) &
    (df['excluFINAL'] == 1)
)
d = df[criterio].copy()
print('N final:', len(d))

print()
print('=== Edad ===')
print(d['edad'].describe().round(2))

print()
print('=== Sexo ===')
print(d['Sexo'].value_counts(normalize=True).round(3) * 100)

print()
print('=== Grado ===')
print(d['grado'].value_counts().sort_index())

print()
print('=== IMC ===')
print(d['imc'].describe().round(2))

print()
print('=== Nivel educativo de los padres (recodificado) ===')
print(d['niveducpadresRECOD'].value_counts(dropna=False))

print()
print('=== Cantidad de personas en el hogar (recodificado) ===')
print(d['cantpersonasviviendaRECOD'].value_counts(dropna=False).sort_index())

N final: 1748

=== Edad ===
count    1748.00
mean       15.44
std         1.59
min        12.00
25%        14.00
50%        15.00
75%        17.00
max        20.00
Name: edad, dtype: float64

=== Sexo ===
Sexo
Masculino    50.0
Femenino     50.0
Name: proportion, dtype: float64

=== Grado ===
grado
1    371
2    379
3    345
4    306
5    276
6     71
Name: count, dtype: int64

=== IMC ===
count    1686.00
mean       20.95
std         2.92
min        12.65
25%        18.94
50%        20.76
75%        22.86
max        29.14
Name: imc, dtype: float64

=== Nivel educativo de los padres (recodificado) ===
niveducpadresRECOD
3.0    1208
2.0     392
1.0      76
NaN      72
Name: count, dtype: int64

=== Cantidad de personas en el hogar (recodificado) ===
cantpersonasviviendaRECOD
1.0     70
2.0    247
3.0    562
4.0    471
5.0    252
6.0     99
7.0     46
NaN      1
Name: count, dtype: int64


El dataset original documenta adolescentes de 12 a 18 años, pero hay de 20.
datoscompletoDEMOG==1 tampoco garantiza demografía 100% completa. Con N=1.748, imc tiene 1.686 casos (62 faltantes) y niveducpadresRECOD tiene 72 faltantes.

In [24]:
criterio = (df['excluidos'] == 0) & (df['edad'] <= 18)
d = df[criterio].copy()

print('Casos con edad > 18:', (d['edad'] > 18).sum())
print(d.loc[d['edad'] > 18, 'edad'].value_counts().sort_index())

Casos con edad > 18: 0
Series([], Name: count, dtype: int64)


Al incluir edad <= 18 en el criterio, la población no contiene casos mayores de 18 años (0)

In [25]:
base = df[df['excluidos'] == 0]
print('Población base (excluidos == 0):', len(base))
print('Casos con edad > 18:', (base['edad'] > 18).sum())
print(base.loc[base['edad'] > 18, 'edad'].value_counts().sort_index())

d = base[base['edad'] <= 18].copy()
print('N final (edad <= 18):', len(d))

Población base (excluidos == 0): 2449
Casos con edad > 18: 24
edad
19.0    20
20.0     4
Name: count, dtype: int64
N final (edad <= 18): 2425


24 casos son el 0.98% de la base

In [ ]:
print('=== Tiempo total de pantalla (h) ===')
print(d['tpopantallatotalhstrimmed'].describe().round(2))

print()
print('=== Por tipo de pantalla (h) ===')
cols_pantalla = ['tpovideojuegoshstrimmed', 'tpojuegotabletcelhstrimmed', 'tporedeshstrimmed', 'tpotvonlinehstrimeed', 'tpootrashstrimed']
print(d[cols_pantalla].describe().round(2).T[['count','mean','std','min','25%','50%','75%','max']])

print()
print('=== Categoría de tiempo total (tpopantallatotalhscat) ===')
print(d['tpopantallatotalhscat'].value_counts(dropna=False).sort_index())

=== Tiempo total de pantalla (h) ===
count    1394.00
mean        6.67
std         2.97
min         0.00
25%         4.33
50%         6.33
75%         9.00
max        13.00
Name: tpopantallatotalhstrimmed, dtype: float64

=== Por tipo de pantalla (h) ===
                             count  mean   std  min   25%  50%  75%    max
tpovideojuegoshstrimmed     1748.0  1.06  1.82  0.0  0.00  0.0  1.5  11.00
tpojuegotabletcelhstrimmed  1748.0  1.84  2.24  0.0  0.17  1.0  2.5  12.50
tporedeshstrimmed           1748.0  2.87  2.51  0.0  1.00  2.0  4.0  12.83
tpotvonlinehstrimeed        1748.0  2.33  1.97  0.0  1.00  2.0  3.0  12.50
tpootrashstrimed            1737.0  0.86  1.70  0.0  0.00  0.0  1.0  12.00

=== Categoría de tiempo total (tpopantallatotalhscat) ===
tpopantallatotalhscat
0.0      3
1.0    471
2.0    571
3.0    349
NaN    354
Name: count, dtype: int64


El tiempo total de pantalla (tpopantallatotalhstrimmed) tiene 354 faltantes, aunque el filtro exige datoscompletoPANTALLAS==1 La palabra "trimmed": el tratamiento de valores extremos no reemplaza el outlier por otro número, lo convierte en NaN. O sea, esos 354 casos no son "faltantes de origen", son valores que los autores del dataset consideraron atípicos/imposibles y los "recortaron".

In [26]:
cols_pantalla = ['tpovideojuegoshstrimmed', 'tpojuegotabletcelhstrimmed', 'tporedeshstrimmed',
                 'tpotvonlinehstrimeed', 'tpootrashstrimed']

print('Nulos por componente:')
print(d[cols_pantalla].isna().sum())
print()
print('Filas con las 5 componentes completas:', d[cols_pantalla].notna().all(axis=1).sum())
print('Filas con algún componente nulo:', d[cols_pantalla].isna().any(axis=1).sum())
print('Total de los autores nulo (para comparar):', d['tpopantallatotalhstrimmed'].isna().sum())

Nulos por componente:
tpovideojuegoshstrimmed        0
tpojuegotabletcelhstrimmed     0
tporedeshstrimmed              0
tpotvonlinehstrimeed           0
tpootrashstrimed              10
dtype: int64

Filas con las 5 componentes completas: 2415
Filas con algún componente nulo: 10
Total de los autores nulo (para comparar): 450


In [27]:
criterio = (df['excluidos'] == 0) & (df['edad'] <= 18)
d = df[criterio].copy()
print('N final:', len(d))

cols_pantalla = ['tpovideojuegoshstrimmed', 'tpojuegotabletcelhstrimmed', 'tporedeshstrimmed',
                 'tpotvonlinehstrimeed', 'tpootrashstrimed']

# Comprobación: ¿el total de los autores es la suma de las 5 partes?
d['suma_componentes'] = d[cols_pantalla].sum(axis=1, skipna=False)
diff = d['suma_componentes'] - d['tpopantallatotalhstrimmed']

print(diff.describe().round(2))
print()
print('Total de los autores nulo:', d['tpopantallatotalhstrimmed'].isna().sum())
print('  de los cuales, componentes completos (superan el recorte de 13 h):',
      ((d['tpopantallatotalhstrimmed'].isna()) & (d['suma_componentes'].notna())).sum())

N final: 2425
count    1975.0
mean        0.0
std         0.0
min         0.0
25%         0.0
50%         0.0
75%         0.0
max         0.0
dtype: float64

Total de los autores nulo: 450
  de los cuales, componentes completos (superan el recorte de 13 h): 440


Hallazgo: el total siempre es exactamente igual a la suma de las 5 partes (diferencia = 0 en los 1.381 casos donde el total existe), y en 337 casos adicionales el total quedó en NaN por el recorte de outliers, pero los 5 componentes individuales sí están disponibles y suman perfecto. conviene recalcular el total propio como suma de los componentes en vez de usar tpopantallatotalhstrimmed tal cual. Eso recupera 337 de los 354 casos que parecían "faltantes" (quedarían solo 17 realmente sin datos)

In [ ]:
criterio = (
    (df['datoscompletoDEMOG'] == 1) &
    (df['datoscompletoSUENO'] == 1) &
    (df['datoscompletoPANTALLAS'] == 1) &
    (df['excluFINAL'] == 1) &
    (df['edad'] <= 18)
)
d = df[criterio].copy()

cols_pantalla = ['tpovideojuegoshstrimmed', 'tpojuegotabletcelhstrimmed', 'tporedeshstrimmed', 'tpotvonlinehstrimeed', 'tpootrashstrimed']
d['pantalla_total_propio'] = d[cols_pantalla].sum(axis=1, skipna=False)

print('Faltantes con el total original:', d['tpopantallatotalhstrimmed'].isna().sum())
print('Faltantes con el total recalculado:', d['pantalla_total_propio'].isna().sum())
print()
print(d['pantalla_total_propio'].describe().round(2))

Faltantes con el total original: 347
Faltantes con el total recalculado: 10

count    1718.00
mean        8.91
std         5.73
min         0.00
25%         4.83
50%         7.33
75%        11.83
max        43.50
Name: pantalla_total_propio, dtype: float64


In [ ]:
criterio = (
    (df['datoscompletoDEMOG'] == 1) &
    (df['datoscompletoSUENO'] == 1) &
    (df['datoscompletoPANTALLAS'] == 1) &
    (df['excluFINAL'] == 1) &
    (df['edad'] <= 18)
)
d = df[criterio].copy()

cols_pantalla = ['tpovideojuegoshstrimmed', 'tpojuegotabletcelhstrimmed', 'tporedeshstrimmed', 'tpotvonlinehstrimeed', 'tpootrashstrimed']
d['pantalla_total_propio'] = d[cols_pantalla].sum(axis=1, skipna=False)

print('Casos con total recalculado > 16h:', (d['pantalla_total_propio'] > 16).sum())
print('Casos con total recalculado > 20h:', (d['pantalla_total_propio'] > 20).sum())
print('Casos con total recalculado > 24h (imposible):', (d['pantalla_total_propio'] > 24).sum())
print()
print(d['pantalla_total_propio'].describe(percentiles=[.9,.95,.99]).round(2))

Casos con total recalculado > 16h: 192
Casos con total recalculado > 20h: 84
Casos con total recalculado > 24h (imposible): 45

count    1718.00
mean        8.91
std         5.73
min         0.00
50%         7.33
90%        16.50
95%        20.00
99%        28.00
max        43.50
Name: pantalla_total_propio, dtype: float64


In [ ]:
criterio = (
    (df['datoscompletoDEMOG'] == 1) &
    (df['datoscompletoSUENO'] == 1) &
    (df['datoscompletoPANTALLAS'] == 1) &
    (df['excluFINAL'] == 1) &
    (df['edad'] <= 18)
)
d = df[criterio].copy()

cols_pantalla = ['tpovideojuegoshstrimmed', 'tpojuegotabletcelhstrimmed', 'tporedeshstrimmed', 'tpotvonlinehstrimeed', 'tpootrashstrimed']
d['suma_componentes'] = d[cols_pantalla].sum(axis=1, skipna=False)

# Total final: original si existe; si no, la suma de componentes, pero solo si es plausible
LIMITE = 16
d['pantalla_total_final'] = d['tpopantallatotalhstrimmed']
usar_suma = d['pantalla_total_final'].isna() & (d['suma_componentes'] <= LIMITE)
d.loc[usar_suma, 'pantalla_total_final'] = d.loc[usar_suma, 'suma_componentes']

print('Faltantes finales:', d['pantalla_total_final'].isna().sum())
print('Casos "rescatados" desde la suma:', usar_suma.sum())
print()
print(d['pantalla_total_final'].describe().round(2))

# ¿El total original mismo tiene casos poco plausibles?
print()
print('Casos del total ORIGINAL > 16h:', (d['tpopantallatotalhstrimmed'] > LIMITE).sum())

Faltantes finales: 202
Casos "rescatados" desde la suma: 145

count    1526.00
mean        7.38
std         3.61
min         0.00
25%         4.50
50%         7.00
75%        10.00
max        16.00
Name: pantalla_total_final, dtype: float64

Casos del total ORIGINAL > 16h: 0


La variable pantalla_total_final con 1.526 de 1.728 casos, sera la variable central del trabajo

**Sueño y somnolencia**

In [ ]:
criterio = (
    (df['datoscompletoDEMOG'] == 1) &
    (df['datoscompletoSUENO'] == 1) &
    (df['datoscompletoPANTALLAS'] == 1) &
    (df['excluFINAL'] == 1) &
    (df['edad'] <= 18)
)
d = df[criterio].copy()

print('=== Duración de sueño, semana (h) ===')
print(d['tposueñosem'].describe().round(2))
print('Faltantes tposueñosem:', d['tposueñosem'].isna().sum())
print('Faltantes tposueñosemTRIMMED:', d['tposueñosemTRIMMED'].isna().sum())

print()
print('=== Duración de sueño, fin de semana (h) ===')
print(d['tpofsueñosem'].describe().round(2))
print('Faltantes tpofsueñosem:', d['tpofsueñosem'].isna().sum())
print('Faltantes tposueñofsemTRIMMED:', d['tposueñofsemTRIMMED'].isna().sum())

print()
print('=== Diferencia de horario de inicio de sueño (semana vs finde) ===')
print(d['difonsetsueño'].describe().round(2))

print()
print('=== Puntaje de somnolencia (Resultado / PDSS) ===')
print(d['Resultado'].describe().round(2))
print()
print('=== Indicador binario de somnolencia (somno01pdss) ===')
print(d['somno01pdss'].value_counts(dropna=False))

=== Duración de sueño, semana (h) ===
count    1728.00
mean        7.16
std         1.20
min         4.00
25%         6.50
50%         7.17
75%         7.92
max        11.83
Name: tposueñosem, dtype: float64
Faltantes tposueñosem: 0
Faltantes tposueñosemTRIMMED: 0

=== Duración de sueño, fin de semana (h) ===
count    1728.00
mean        9.42
std         2.08
min         1.00
25%         8.00
50%         9.50
75%        11.00
max        14.00
Name: tpofsueñosem, dtype: float64
Faltantes tpofsueñosem: 0
Faltantes tposueñofsemTRIMMED: 0

=== Diferencia de horario de inicio de sueño (semana vs finde) ===
count    1728.00
mean        3.05
std         3.49
min         0.00
25%         1.00
50%         2.00
75%         3.50
max        17.17
Name: difonsetsueño, dtype: float64

=== Puntaje de somnolencia (Resultado / PDSS) ===
count    1728.00
mean       18.18
std         5.13
min         2.00
25%        15.00
50%        18.00
75%        22.00
max        31.00
Name: Resultado, dtype: float64


El flag datoscompletoSUENO==1 tiene cero faltantes en las variables de sueño. Pero difonsetsueño (diferencia de horario de inicio de sueño entre semana y finde) llega a un máximo de 17,17 horas, que es un valor extraño para una diferencia de horario — más de medio día de diferencia no es creíble

In [ ]:
criterio = (
    (df['datoscompletoDEMOG'] == 1) &
    (df['datoscompletoSUENO'] == 1) &
    (df['datoscompletoPANTALLAS'] == 1) &
    (df['excluFINAL'] == 1) &
    (df['edad'] <= 18)
)
d = df[criterio].copy()

print('Distribución de difonsetsueño en los extremos:')
print(d['difonsetsueño'].describe(percentiles=[.9,.95,.99]).round(2))
print()
print('Casos con difonsetsueño > 8h:', (d['difonsetsueño'] > 8).sum())

# Comparar contra el cálculo directo desde las horas de acostarse
cols_horas = ['horasueñosemanahr', 'horaacostarfsemanahr']
print()
print(d[cols_horas].describe().round(2))

Distribución de difonsetsueño en los extremos:
count    1728.00
mean        3.05
std         3.49
min         0.00
50%         2.00
90%         6.50
95%        13.32
99%        15.50
max        17.17
Name: difonsetsueño, dtype: float64

Casos con difonsetsueño > 8h: 137

       horasueñosemanahr  horaacostarfsemanahr
count            1728.00               1728.00
mean               23.15                 25.17
std                 1.12                  2.06
min                19.50                 19.00
25%                22.50                 23.83
50%                23.50                 25.00
75%                23.50                 26.50
max                30.50                 30.83


Esto confirma que hay un grupo de casos calculados mal, las horas base (horasueñosemanahr media 23,15 = ~23:09, horaacostarfsemanahr media 25,17 = ~1:10 AM) ya vienen en formato extendido (pasan de 24 para indicar que cruzan la medianoche), así que la diferencia directa entre esas dos columnas debería alcanzar sin necesidad de ningún ajuste circular.

In [ ]:
criterio = (
    (df['datoscompletoDEMOG'] == 1) &
    (df['datoscompletoSUENO'] == 1) &
    (df['datoscompletoPANTALLAS'] == 1) &
    (df['excluFINAL'] == 1) &
    (df['edad'] <= 18)
)
d = df[criterio].copy()

# Diferencia recalculada directamente (ambas columnas ya están en formato extendido >24h)
d['diff_recalculada'] = (d['horaacostarfsemanahr'] - d['horasueñosemanahr']).abs()

print(d['diff_recalculada'].describe(percentiles=[.9,.95,.99]).round(2))
print()

# Comparar contra la columna original del dataset
comparacion = (d['diff_recalculada'] - d['difonsetsueño']).abs()
print('Diferencia entre mi cálculo y la columna original:')
print(comparacion.describe().round(2))
print()
print('Casos donde coinciden (diferencia < 0.1h):', (comparacion < 0.1).sum())
print('Casos donde NO coinciden:', (comparacion >= 0.1).sum())

count    1728.00
mean        2.18
std         1.68
min         0.00
50%         1.83
90%         4.50
95%         5.50
99%         7.00
max         9.50
Name: diff_recalculada, dtype: float64

Diferencia entre mi cálculo y la columna original:
count    1728.00
mean        0.87
std         3.09
min         0.00
25%         0.00
50%         0.00
75%         0.00
max        12.00
dtype: float64

Casos donde coinciden (diferencia < 0.1h): 1596
Casos donde NO coinciden: 132


diff_recalculada es mucho más creíble (máximo 9,5h, sin el salto raro del percentil 95), y en 1.596 de 1.728 casos (92,4%) coincide exactamente con la columna original del dataset — eso valida que el cálculo directo es correcto. Los 132 casos donde no coinciden son justamente los que tenían el problema. **Decisión** : descartar difonsetsueño tal como viene en el dataset y usar diff_recalculada (a partir de horasueñosemanahr y horaacostarfsemanahr), debido a la inconsistencia en la variable original en el 7,6% de los casos.

**Relación entre pantallas y sueño/somnolencia**

In [ ]:
criterio = (
    (df['datoscompletoDEMOG'] == 1) &
    (df['datoscompletoSUENO'] == 1) &
    (df['datoscompletoPANTALLAS'] == 1) &
    (df['excluFINAL'] == 1) &
    (df['edad'] <= 18)
)
d = df[criterio].copy()

cols_pantalla = ['tpovideojuegoshstrimmed', 'tpojuegotabletcelhstrimmed', 'tporedeshstrimmed', 'tpotvonlinehstrimeed', 'tpootrashstrimed']
d['suma_componentes'] = d[cols_pantalla].sum(axis=1, skipna=False)
d['pantalla_total_final'] = d['tpopantallatotalhstrimmed']
usar_suma = d['pantalla_total_final'].isna() & (d['suma_componentes'] <= 16)
d.loc[usar_suma, 'pantalla_total_final'] = d.loc[usar_suma, 'suma_componentes']

cols_numericas = ['pantalla_total_final'] + cols_pantalla + ['tposueñosem', 'tpofsueñosem', 'Resultado']
print(d[cols_numericas].corr(method='spearman').round(2)['pantalla_total_final'])

print()
print('=== Sueño en semana según nivel de somnolencia ===')
print(d.groupby('somno01pdss')['tposueñosem'].describe()[['count','mean','std']].round(2))

pantalla_total_final          1.00
tpovideojuegoshstrimmed       0.27
tpojuegotabletcelhstrimmed    0.45
tporedeshstrimmed             0.51
tpotvonlinehstrimeed          0.49
tpootrashstrimed              0.27
tposueñosem                  -0.11
tpofsueñosem                 -0.02
Resultado                     0.09
Name: pantalla_total_final, dtype: float64

=== Sueño en semana según nivel de somnolencia ===
              count  mean   std
somno01pdss                    
0             530.0  7.38  1.22
1            1198.0  7.06  1.17


El hecho de que el tiempo total de pantalla, tomado como un solo número, no prediga bien sueño ni somnolencia confirma  el argumento que se plantea en el informe: "si todos los adolescentes son tratados como un único grupo, se pierde información".El enfoque agregado no alcanza y por eso se justifica buscar subgrupos con composiciones de uso distintas (no solo "cuánto" sino "en qué" usan la pantalla). El clustering va a permitir caracterizar perfiles distintos, sin garantizar que esos perfiles se diferencien fuertemente en sueño

**Cierre de EDA**

## Conclusiones del Análisis Exploratorio de Datos (EDA)

**Fuente y estructura del archivo**

El archivo `repositorio.xlsx` contiene 4.115 registros y 125 columnas, sin filas ni columnas duplicadas. El estudio original documenta 1.257 adolescentes, por lo que el archivo disponible es una versión ampliada del dataset publicado.

**Definición de la muestra válida**

El dataset incluye indicadores propios de completitud (`datoscompletoDEMOG`, `datoscompletoSUENO`, `datoscompletoPANTALLAS`, `excluFINAL`). Se verificó que `excluFINAL == 1` (3.251 casos) no garantiza por sí solo la completitud de las variables demográficas: dentro de ese subconjunto, `edad` sigue faltando en 399 casos. Por lo tanto, se definió un criterio propio de inclusión: los tres indicadores de completitud en 1, `excluFINAL == 1` y edad dentro del rango documentado por el estudio (≤18 años). Esto deja una muestra final de **1.728 adolescentes**.

Se evaluó el sesgo de selección comparando el grupo incluido contra el excluido en variables clave: no se encontraron diferencias relevantes en edad (15,44 vs. 15,75 años) ni en tiempo total de pantalla (6,67 vs. 6,75 h). Sí se observó una diferencia en la composición por sexo: el grupo incluido queda balanceado 50/50, mientras que el excluido está sesgado hacia el sexo femenino (59,3%), lo que sugiere que las respuestas incompletas se concentran algo más en ese grupo.

**Tratamiento de outliers y variables reconstruidas**

- *Tiempo total de pantalla:* la variable `tpopantallatotalhstrimmed` tiene un 27,7% de faltantes, producto del tratamiento de valores extremos aplicado por los autores (los outliers se convierten en NaN en lugar de recortarse a un valor). Se verificó que el total es exactamente igual a la suma de sus 5 componentes en todos los casos donde ambos existen, por lo que se reconstruyó una variable `pantalla_total_final` completando los faltantes con la suma de componentes, pero solo cuando esa suma es ≤16 horas (límite de plausibilidad, validado contra el hecho de que el total original nunca supera ese valor). Esto redujo los faltantes de 347 a 202 casos.
- *Diferencia de horario de inicio de sueño:* la variable `difonsetsueño` presentaba valores extremos poco creíbles (hasta 17,17 h de diferencia). Se recalculó directamente como la diferencia absoluta entre `horaacostarfsemanahr` y `horasueñosemanahr`, obteniendo una distribución consistente (máximo 9,5 h) que coincide con la variable original en el 92,4% de los casos, confirmando un error de cálculo en el 7,6% restante.

**Variables descartadas**

Se identificaron variables con niveles de faltantes que las vuelven inviables para el análisis: `WAZ` (99,3% de faltantes) y las variables de rendimiento académico `prom`, `aplazos`, `matematica` y `lengua` (~69% de faltantes cada una).

**Relación entre uso de pantallas y sueño/somnolencia**

Las correlaciones (Spearman) entre el tiempo total de pantalla y las variables de sueño son débiles: -0,11 con horas de sueño en semana, -0,02 con horas de sueño en fin de semana, y 0,09 con el puntaje de somnolencia (PDSS). La diferencia en horas de sueño entre adolescentes con y sin somnolencia positiva es de apenas 19 minutos (7,06 h vs. 7,38 h). Esta ausencia de relación lineal fuerte entre el tiempo *total* de pantalla y el sueño respalda empíricamente el enfoque del trabajo: un análisis agregado no captura diferencias relevantes, lo que justifica buscar perfiles diferenciados según la composición del uso de pantallas mediante clustering, en lugar de predecir a partir de un único indicador. Cabe aclarar que la correlación de Spearman solo captura relaciones monótonas entre el tiempo total de pantalla y el sueño; no descarta que existan patrones diferenciados según la combinación de actividades (redes, videojuegos, TV), que es justamente lo que el enfoque de clustering busca explorar.

**Dataset resultante para el modelado**

Muestra final: 1.728 adolescentes, con variables demográficas, tiempo de pantalla por tipo de actividad (`pantalla_total_final` y sus 5 componentes), variables de sueño y somnolencia sin faltantes, y variables de contexto del hogar con faltantes menores pendientes de resolver (IMC: 62 casos; nivel educativo de los padres: 72 casos).